### PR TITLE
Multisite revision

### PR DESCRIPTION
**Pull request by DominikaK**
| Question      | Answer
| ------------- | ---
| JIRA Ticket   | https://issues.ibexa.co/browse/IBX-493
| Versions      | 3.3+

### CODE REVIEW

File: docs/guide/multisite/siteaccess_aware_configuration.md [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+- `scope` is the SiteAccess, SiteAccess group, `default` or `global`
+- `parameter_name` is the actual setting *identifier*
+
+For more information about the ConfigResolver, namespaces and scopes, see [[[= product_name =]] configuration basics](../configuration.md).
+
+### Semantic configuration parsing
+
+To parse semantic configuration, create a `Configuration` class which extends
+`eZ\Bundle\EzPublishCoreBundle\DependencyInjection\Configuration\SiteAccessAware\Configuration`
+and the extend its `generateScopeBaseNode()` method:
```
Comment (alongosz) [inline docs/guide/multisite/siteaccess_aware_configuration.md:24]: Not sure where to place this remark, so I'll put it here. While this is correct, it's not nearly enough. There's an implicit assumption of the existence of AppBundle which comes from Symfony 2- 3 boilerplate code. This is no longer the case for Symfony 5+ - bundles are gone. On the other hand Bundle is required to load configuration extension, so need to be added.
One of the possibilities of full configuration is shown is [this diff](https://github.com/alongosz/ibexa-open-source/compare/564a5cf..e9887ab)

**Review by alongosz — CHANGES_REQUESTED** [review]
Review for `docs/guide/multisite/siteaccess_aware_configuration.md` only.
I pointed out several places in code which need to be changed, but that's not enough. I would recommend a rewrite using [this diff](https://github.com/alongosz/ibexa-open-source/compare/564a5cf..e9887ab).

Full example: https://github.com/alongosz/ibexa-open-source/tree/ibx-493-multisite-revision

File: docs/guide/multisite/injecting_siteaccess.md [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    public function __construct(ContentService $contentService)
+    {
+        $this->contentService = $contentService;
+    }
+
+    public function setSiteAccess(SiteAccess $siteAccess = null)
+    {
+        $this->siteAccess = $siteAccess;
+    }
+}
```
Comment (alongosz) [inline docs/guide/multisite/injecting_siteaccess.md:45]: Following the previous comment:
```suggestion
declare(strict_types=1);

namespace App;

use eZ\Publish\API\Repository\ContentService;
use eZ\Publish\Core\MVC\Symfony\SiteAccess\SiteAccessServiceInterface;

class MyService
{
    /** @var \eZ\Publish\API\Repository\ContentService */
    private $contentService;

    /** @var \eZ\Publish\Core\MVC\Symfony\SiteAccess\SiteAccessServiceInterface */
    private $siteAccessService;

    public function __construct(
        SiteAccessServiceInterface $siteAccessService,
        ContentService $contentService
    ) {
        $this->siteAccessService = $siteAccessService;
        $this->contentService = $contentService;
    }
}
```

File: docs/guide/multisite/multisite_configuration.md [hunk 3]
```diff
@@ -0,0 +1,193 @@
+# Multisite configuration
+
+You configure the available SiteAccesses and under the `ezplatform.siteaccess` key.
```
Comment (alongosz) [inline docs/guide/multisite/multisite_configuration.md:3]: This "and" in the middle feels like a typo.

File: docs/guide/multisite/multisite_configuration.md [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+A SiteAccess can be part of several groups. SiteAccess configuration has always precedence over group configuration.
+
+#### `admin` SiteAccess
+
+The predefined `admin` SiteAccess in `admin_group` (configured in `config/packages/ezplatform_admin_ui.yaml`) serves the Back Office.
+Do not remove this group.
+If you need a multisite setup with multiple Back Offices, add any additional administration SiteAccesses to `admin_group`.
+
+In cases where the sites are on separate databases, each needs its own [repository](../configuration.md#configuration-examples)
+(including their own storage and search connection), var dir, [cache pool](../persistence_cache.md#persistence-cache-configuration),
```
Comment (alongosz) [inline docs/guide/multisite/multisite_configuration.md:42]: I think it needs to be stated more clearly, maybe as a next sentence, emphasized as caution/warning that two different SiteAccesses can have different `var_dir` only if they also have different `repository`. 

_This is a flaw in the architecture - `var_dir` should be a `repository`, not SA setting, but we can't change that easily._

File: docs/guide/multisite/multisite_configuration.md [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+                siteaccess_list: [site, de, fr, no]
+        de:
+            page_builder:
+                siteaccess_list: [site, de]
+```
+
+If you are using multiple domains, list all domains for an admin SiteAccess under `siteaccess_hosts`:
+
+``` yaml
+ezpublish:
```
Comment (alongosz) [inline docs/guide/multisite/multisite_configuration.md:147]:
```suggestion
ezplatform:
```

File: docs/guide/multisite/site_factory.md [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+![Site Factory enabled](../img/site_factory_site_list.png "Site Factory enabled")
+
+### Define site directory
+
+You can adjust the place where the directory of the new site is created (Location with ID 2 by default).
+To do it, go to `config/packages/ezplatform_site_factory.yaml`, and add the following parameter:
+
+``` yaml
+parameters:
+    ezsettings.default.site_factory.sites_location_id: 42
```
Comment (alongosz) [inline docs/guide/multisite/site_factory.md:159]: @mikadamczyk is this really a parameter or is there real semantic config for that?

**Review by alongosz — CHANGES_REQUESTED** [review]
Site Factory section should be reviewed by @mikadamczyk 
+1 for `docs/guide/multisite/siteaccess_aware_configuration.md`
Remarks for the other parts:

File: docs/guide/multisite/injecting_siteaccess.md [hunk 7]
```diff
@@ -0,0 +1,39 @@
+# Injecting SiteAccess
+
+The [service container](../service_container.md) exposes the SiteAccess as the `@ezpublish.siteaccess` service,
+so you can inject it using setter injection into any custom service.
+The service should implement the `eZ\Publish\Core\MVC\Symfony\SiteAccess\SiteAccessAware` interface.
+
+For example, define a service which depends on the Repository's ContentService and the current SiteAccess.
```
Comment (alongosz) [inline docs/guide/multisite/injecting_siteaccess.md:7]:
```suggestion
The [service container](../service_container.md) exposes the SiteAccess through the `@ezpublish.siteaccess_service` service, which fulfills the `\eZ\Publish\Core\MVC\Symfony\SiteAccess\SiteAccessServiceInterface` contract, 
so you can inject it into any custom service constructor, type hinting the mentioned contract. Current SiteAccess can be obtained from that service by calling `SiteAccessServiceInterface::getCurrent` method.

For example, define a service which depends on the Repository's ContentService and the SiteAccessService.
```

**Review by alongosz — CHANGES_REQUESTED** [review]

File: docs/guide/multisite/injecting_siteaccess.md [hunk 8]
```diff
@@ -0,0 +1,39 @@
+# Injecting SiteAccess
+
+The [service container](../service_container.md) exposes the SiteAccess as the `@ezpublish.siteaccess` service,
```
Comment (DominikaK) [inline docs/guide/multisite/injecting_siteaccess.md:3]:
```suggestion
The [service container](../service_container.md) exposes the SiteAccess as the `@ezpublish.siteaccess` service,
```

File: docs/guide/multisite/injecting_siteaccess.md [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+so you can inject it using setter injection into any custom service.
+The service should implement the `eZ\Publish\Core\MVC\Symfony\SiteAccess\SiteAccessAware` interface.
+
+For example, define a service which depends on the Repository's ContentService and the current SiteAccess.
+
```
Comment (DominikaK) [inline docs/guide/multisite/injecting_siteaccess.md:8]:
```suggestion
The [service container](../service_container.md) exposes the SiteAccess through the `@ezpublish.siteaccess_service` service, which fulfills the `\eZ\Publish\Core\MVC\Symfony\SiteAccess\SiteAccessServiceInterface` contract.
This means you can inject it into any custom service constructor, type hinting that contract.
You can get the current SiteAccess from that service by calling the `SiteAccessServiceInterface::getCurrent` method.
	
For example, define a service which depends on the Repository's ContentService and the SiteAccessService.

```

**Review by alongosz — APPROVED** [review]
+1 for non-Site Factory changes

File: docs/guide/multisite/site_factory.md (hunk 6 above)
Comment (mikadamczyk) [reply docs/guide/multisite/site_factory.md:159]: Hi @alongosz! It is a real semantic configuration https://github.com/ezsystems/ezplatform-site-factory/blob/master/src/bundle/DependencyInjection/Parser/SiteFactory.php#L35

**Review by mikadamczyk — APPROVED** [review]

File: docs/guide/multisite/site_factory.md (hunk 6 above)
Comment (DominikaK) [reply docs/guide/multisite/site_factory.md:159]: @mikadamczyk Shouldn't I then replace it with 'regular' config?
```
ezplatform:
    system:
        default:
            site_factory:
                sites_location_id: 2
```

Comment (alongosz) [reply docs/guide/multisite/site_factory.md]: It definitely cannot be documented as a parameter in this case.

Comment (DominikaK) [reply docs/guide/multisite/site_factory.md]: Yep, already replaced with the example above.

**Review by juskora — CHANGES_REQUESTED** [review]

File: docs/guide/multisite/multisite.md [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+
+A SiteAccess is a set of configuration settings that the application uses when you access the site through a specific address.
+When the user visits the site, the system analyzes the URI and compares it to rules specified in the configuration.
+If it finds a set of fitting rules, this SiteAccess is used.
+
+You can create a multisite setup with SiteAccesses covering different use cases, for example:
+
+- [different language versions of the webpage](set_up_translation_siteaccess.md)
+- [a special sub-site for a campaign or event](set_up_campaign_siteaccess.md)
+- additional Back Office
```
[the remark this answers is not in the document]
Comment (DominikaK) [reply docs/guide/multisite/multisite.md:13]:
```suggestion
- additional [Back Office](multisite_configuration.md#admin-siteaccess)
```

File: docs/guide/multisite/site_factory.md [hunk 11]
```diff
@@ -0,0 +1,219 @@
...
+## Disable Site Factory
+
+Enabled Site Factory may cause following performance issues:
+
+- [ConfigResolver](../config_dynamic.md#configresolver) looks for SiteAccesses in the database
+- Site Factory matchers are connected to the database in search for new SiteAccesses
+
+You can disable Site Factory to boost ConfigResolver performance.
+Keep in mind that with disabled Site Factory you are unable to add new sites or use existing ones.
+
```
[the remark this answers is not in the document]
Comment (DominikaK) [reply docs/guide/multisite/site_factory.md:192]: To me a note implies side information, something not in the direct flow of the text, and this inability results directly from disabling SF.

File: docs/guide/multisite/siteaccess_aware_configuration.md (hunk 1 above)
Comment (DominikaK) [reply docs/guide/multisite/siteaccess_aware_configuration.md:31]:
```suggestion
To parse semantic configuration, create a `Configuration` class which extends
`eZ\Bundle\EzPublishCoreBundle\DependencyInjection\Configuration\SiteAccessAware\Configuration`
and extend its `generateScopeBaseNode()` method:
```

Comment (DominikaK) [reply docs/guide/multisite/siteaccess_aware_configuration.md:31]:
```suggestion
To parse semantic configuration, create a `Configuration` class which extends
`eZ\Bundle\EzPublishCoreBundle\DependencyInjection\Configuration\SiteAccessAware\Configuration`
and then extend its `generateScopeBaseNode()` method:
```

File: docs/guide/multisite/siteaccess_aware_configuration.md [hunk 12]
```diff
@@ -0,0 +1,261 @@
...
+            custom_setting:
+                string: value
+                number: 123
+                enabled: false
+```
+
+### Mapping to internal settings
+
+Semantic configuration must always be mapped to internal key/value settings within the service container.
+You usually do it the service container extension.
```
[the remark this answers is not in the document]
Comment (DominikaK) [reply docs/guide/multisite/siteaccess_aware_configuration.md:91]:
```suggestion
You usually do it in the [service container](../../api/service_container.md) extension.
```

**Review by juskora — APPROVED** [review]

**Pull request merged**